### PR TITLE
buildGraalvmNativeImage: init

### LINKED ISSUE #140544: GraalVM native images unnecessarily reference the GraalVM store path
### Describe the bug

When building a native image using `graalvm11-ce`, the produced (dynamically linked) executable will be linked against libraries in `"${graalvm11-ce.out}/lib"`, causing a runtime dependency on the entire store path. However this directory only contains symlinks to other packages.

By teaching GraalVM how to link against those libraries *directly* we would instantly save ~1GB of closure size for every dynamically linked binary produced by GraalVM since the runtime dependency would be eliminated. 

The question is of course *how* doable this is. Looking at the derivation, it seems that the symlinks were introduced in the first place to counter how inflexible GraalVM is in this regard.

### Notify maintainers

@hlolli @glittershark @babariviere @ericdallo

### PR DESCRIPTION
**Pull request by thiagokokada**
###### Motivation for this change

We have many packages using GraalVM's `native-image` in nixpkgs. However, building it wrong can result in huge closures (see https://github.com/NixOS/nixpkgs/pull/140287 and https://github.com/NixOS/nixpkgs/issues/140544).

So this PR introduces `buildGraalvmNativeImage` builder, making it easier to package new derivations, and specially, making it easier to package them right.

For now it is a simple builder that supports only one case: 1 Jar -> 1 Executable. However this should be the vast majority of the cases for a GraalVM case, and we can always expand it in the future to support new cases.

Fixes #140544.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (thiagokokada) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:20]: `extraNativeImageBuildArgs` doesn't expand shell variables, this is why I had to hardcode it to `/tmp` instead.

File: pkgs/build-support/build-graalvm-native-image/default.nix
Comment (leo60228) [inline pkgs/build-support/build-graalvm-native-image/default.nix:14]: Is there a reason for using GraalVM 11 instead of 17? Both are LTS, I believe.

File: pkgs/build-support/build-graalvm-native-image/default.nix
Comment (leo60228) [inline pkgs/build-support/build-graalvm-native-image/default.nix:49]: I'm pretty sure this will completely overwrite `args.meta`, not merge with it.

File: pkgs/development/interpreters/clojure/babashka.nix
Comment (leo60228) [inline pkgs/development/interpreters/clojure/babashka.nix:16]: Does it make sense for `--verbose` to be here?

File: pkgs/top-level/all-packages.nix
Comment (leo60228) [inline pkgs/top-level/all-packages.nix:12345]: I guess it's consistent with the `graalvmXX-ce` packages, but `buildGraalvmNativeImage` looks weird to me compared to `buildGraalVMNativeImage`.

File: pkgs/build-support/build-graalvm-native-image/default.nix
Comment (thiagokokada) [reply pkgs/build-support/build-graalvm-native-image/default.nix:14]: I could set this to 17, but them I would set need to set GraalVM 11 to all packages.

I don't think I ever saw a package building with GraalVM 17 instead of 11 yet.

File: pkgs/top-level/all-packages.nix
Comment (thiagokokada) [reply pkgs/top-level/all-packages.nix:12345]: Yeah, but I prefer to be consistent here.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 147136` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

File: pkgs/build-support/build-graalvm-native-image/default.nix
Comment (leo60228) [reply pkgs/build-support/build-graalvm-native-image/default.nix:14]: That's fair.

File: pkgs/top-level/all-packages.nix
Comment (leo60228) [reply pkgs/top-level/all-packages.nix:12345]: That's true. Neither option is that great here, really.

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 147136` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (ericdallo) [inline pkgs/development/tools/misc/clojure-lsp/default.nix:24]: Maybe we should consider moving this one to the common `nativeImageBuildArgs`  as if you are using native-image you pretty much don't want a fallback image as result of the compilation?

**Review by ericdallo — APPROVED** [review]
Well done, really good refactor!

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (thiagokokada) [reply pkgs/development/tools/misc/clojure-lsp/default.nix:24]: I don't think so, we are not statically linking here so building without `--no-fallback` should be valid too.

**Comment by thiagokokada** [comment]
I would like one more review here before merging.

CC @hlolli @babariviere @glittershark @bennyandresen @glittershark.

File: pkgs/development/tools/misc/clojure-lsp/default.nix
Comment (ericdallo) [reply pkgs/development/tools/misc/clojure-lsp/default.nix:24]: yeah, I think so, is just that the fallback image is a executable that just run the jar with java 🤷🏻 , but it's still valid indeed

**Comment by thiagokokada** [comment]
Result of `nixpkgs-review pr 147136` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>7 packages built:</summary>
  <ul>
    <li>babashka</li>
    <li>clj-kondo</li>
    <li>clojure-lsp</li>
    <li>jet</li>
    <li>neil</li>
    <li>vscode-extensions.betterthantomorrow.calva</li>
    <li>zprint</li>
  </ul>
</details>

**Review by ericdallo — APPROVED** [review]

**Comment by thiagokokada** [comment]
Going to merge this PR (once the tests pass), any further improvements can go on other PRs.

**Pull request merged**
